Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -332,3 +391,132 @@ private func sandboxProfile() -> String {
+final class ManifestLoadRule: LLBuildRule {
+
+    struct RuleKey: LLBuildKey {
+        typealias BuildValue = RuleValue
+        typealias BuildRule = ManifestLoadRule
+
+        let path: AbsolutePath
+        let manifestVersion: ManifestVersion
+    }
+
+    struct RuleValue: LLBuildValue {
+        let result: Result<String, StringError>
+    }
+
+    override class var ruleName: String { return "\(ManifestLoadRule.self)" }
+
+    private let key: RuleKey
+    private weak var loader: ManifestLoader!
+
+    init(_ key: Key, loader: ManifestLoader) {
+        self.key = RuleKey(key)
+        self.loader = loader
+        super.init()
+    }
+
+    override func start(_ engine: LLTaskBuildEngine) {
+        engine.taskNeedsInput(SwiftPMVersionRule.RuleKey(), inputID: 1)
+        engine.taskNeedsInput(FileInfoRule.RuleKey(path: key.path), inputID: 2)
+    }
+
+    override func inputsAvailable(_ engine: LLTaskBuildEngine) {
+        let value = RuleValue(result: Result(string: {
+            try loader.parse(path: key.path, manifestVersion: key.manifestVersion)
+        }))
+        engine.taskIsComplete(value)
+    }
+}
+
+/// A rule to get file info of a file on disk.
+// FIXME: Find a proper place for this rule.
+final class FileInfoRule: LLBuildRule {
+
+    struct RuleKey: LLBuildKey {
+        typealias BuildValue = RuleValue
+        typealias BuildRule = FileInfoRule
+
+        let path: AbsolutePath
+    }
+
+    struct RuleValue: LLBuildValue, Equatable {
+        let fileInfo: Result<FileInfo, StringError>
+    }
+
+    override class var ruleName: String { return "\(FileInfoRule.self)" }
+
+    private let key: RuleKey
+
+    init(_ key: Key) {
+        self.key = RuleKey(key)
+        super.init()
+    }
+
+    override func isResultValid(_ priorValue: Value) -> Bool {
+        return getFileInfo(key.path) == RuleValue(priorValue)
+    }
+
+    override func inputsAvailable(_ engine: LLTaskBuildEngine) {
+        engine.taskIsComplete(getFileInfo(key.path))
+    }
+
+    private func getFileInfo(_ path: AbsolutePath) -> RuleValue {
+        return RuleValue(fileInfo: Result(string: {
+            try localFileSystem.getFileInfo(key.path)
+        }))
+    }
+}
+
+/// A rule to compute the current version of the pacakge manager.
+///
+/// This rule will always run.
+// FIXME: Find a proper place for this rule.
+final class SwiftPMVersionRule: LLBuildRule {
+
+    struct RuleKey: LLBuildKey {
+        typealias BuildValue = RuleValue
+        typealias BuildRule = SwiftPMVersionRule
+    }
+
+    struct RuleValue: LLBuildValue, Equatable {
+        let version: String
+    }
+
+    override class var ruleName: String { return "\(SwiftPMVersionRule.self)" }
+
+    override func isResultValid(_ priorValue: Value) -> Bool {
+        // Always rebuild this rule.
+        return false
+    }
+
+    override func inputsAvailable(_ engine: LLTaskBuildEngine) {
+        let version = Versioning.currentVersion.displayString
```

Comment:
Bear in mind this should probably be derived from the actual swiftpm binary, not just its own version.

Reply:
This string is something like swiftpm 4.2, right? My point was it won't cause a rebuild if you just land a new PR in SwiftPM that causes the logic to change, for that something like the timestamp or hash of swift-build is more reliable (and also more annoying)